File: python/program_generation.py
from typing import Callable, Union

Ident = str

class Envcomp:
    pass

class Var(Envcomp):
    def __init__(self, x: Ident):
        self.x = x

class Lam(Envcomp):
    def __init__(self, x: Ident, e: Envcomp):
        self.x = x
        self.e = e

class App(Envcomp):
    def __init__(self, f: Envcomp, a: Envcomp):
        self.f = f
        self.a = a

class LamD(Envcomp):
    def __init__(self, x: Ident, e: Envcomp):
        self.x = x
        self.e = e

class AppD(Envcomp):
    def __init__(self, f: Envcomp, a: Envcomp):
        self.f = f
        self.a = a

Exp = Union[Var, Lam, App, LamD, AppD]

class Fun:
    def __init__(self, f: Callable[[Value], Value]):
        self.f = f

class Code:
    def __init__(self, e: Exp):
        self.e = e

Value = Union[Fun, Code]

def spec(e: Exp, r: dict[Ident, Value]) -> Value:
    if isinstance(e, Var):
        return r.get(e.x, Code(e))
    elif isinstance(e, Lam):
        return Fun(lambda y: spec(e.e, {**r, e.x: y}))
    elif isinstance(e, App):
        f, a = spec(e.f, r), spec(e.a, r)
        if isinstance(f, Fun):
            return f.f(a)
        else:
            raise TypeError("Type error")
    elif isinstance(e, LamD):
        xx = gensym()
        r[e.x] = Code(Var(xx))
        body = spec(e.e, r)
        if isinstance(body, Code):
            return Code(Lam(xx, body.e))
        else:
            raise TypeError("Type error")
    elif isinstance(e, AppD):
        f, a = spec(e.f, r), spec(e.a, r)
        if isinstance(f, Code) and isinstance(a, Code):
            return Code(App(f.e, a.e))
        else:
            raise TypeError("Type error")

File: go/program_generation.go
package combinators

import "fmt"

type Ident = string

type Envcomp interface{}

type Var struct {
    x Ident
}

type Lam struct {
    x Ident
    e Envcomp
}

type App struct {
    f, a Envcomp
}

type LamD struct {
    x Ident
    e Envcomp
}

type AppD struct {
    f, a Envcomp
}

type Exp interface{}

type Var' struct {
    x Ident
}

type Lam' struct {
    x Ident
    e Exp
}

type App' struct {
    f, a Exp
}

type LamD' struct {
    x Ident
    e Exp
}

type AppD' struct {
    f, a Exp
}

type Value interface{}

type Fun struct {
    f func(Value) Value
}

type Code struct {
    e Exp
}

func spec(e Exp, r map[Ident]Value) Value {
    switch exp := e.(type) {
    case Var':
        if v, ok := r[exp.x]; ok {
            return v
        } else {
            return Code{Var'{exp.x}}
        }
    case Lam':
        return Fun{func(y Value) Value {
            r[exp.x] = y
            return spec(exp.e, r)
        }}
    case App':
        f, a := spec(exp.f, r), spec(exp.a, r)
        if ff, ok := f.(Fun); ok {
            return ff.f(a)
        } else {
            panic("Type error")
        }
    case LamD':
        xx := gensym()
        r[exp.x] = Code{Var'{xx}}
        body := spec(exp.e, r)
        return Code{Lam'{xx, body.(Code).e}}
    case AppD':
        f, a := spec(exp.f, r), spec(exp.a, r)
        ff, ok1 := f.(Code)
        aa, ok2 := a.(Code)
        if ok1 && ok2 {
            return Code{App'{ff.e, aa.e}}
        } else {
            panic("Type error")
        }
    default:
        panic("Invalid expression")
    }
}